### ISSUE TITLE
feat: Epictetus as Marcus's Loss Function — Closing the Self-Improvement Loop

### ISSUE THREAD

**Issue by lwgray** [lwgray/marcus#257]
## The Big Picture

Marcus has the bones of a self-improving multi-agent system, but the pieces aren't connected yet. This issue maps how they fit together — and identifies what's missing to close the loop.

The three Stoics form a triangle:

```
            Marcus (coordination)
           ╱                     ╲
      forward pass            weight update
         ╱                         ╲
   Cato (observation) ──────── Epictetus (evaluation)
            backward pass
```

- **Marcus** runs the experiment (forward pass — tasks flow through decomposition → assignment → execution → code)
- **Cato** observes everything and aggregates it into unified `ProjectHistory` (the data layer that makes attribution possible)
- **Epictetus** evaluates the output AND reads Cato's aggregated history to understand *process*, not just *product* (the loss function)
- **Blame attribution** traces Epictetus findings back through Cato's timeline to the specific instruction envelope that caused them (backward pass)
- **The optimizer** feeds corrections into Marcus's learning system parameters (weight update)

### The Backpropagation Analogy

| Neural Net | Marcus Equivalent | Where It Lives | Status |
|---|---|---|---|
| **Parameters/Weights** | PatternLearner, ContextualLearner, Memory agent profiles, AdaptiveDependencyInferer weights | `src/learning/`, `src/ai/learning/`, `src/core/memory.py`, `src/core/adaptive_dependencies.py` | Built but disconnected (#252) |
| **Forward Pass** | Experiment execution: decompose → assign → agents execute → code | `src/marcus_mcp/`, `src/core/`, `skills/marcus` | Built |
| **Data Recording** | Cato's observation layer: events, conversations, decisions, artifacts, snapshots → `ProjectHistory` | `src/analysis/aggregator.py`, `src/core/project_history.py`, `src/logging/`, `src/monitoring/` | Built |
| **Loss Function** | Epictetus audit: 9-dimension rubric, structured JSON report, quality deficit from 5.0 | `skills/epictetus/` | Built (skill only) |
| **Backward Pass** | Blame attribution: trace each defect through Cato's timeline to system/agent/prompt layer | Designed in #255, not built | **Missing** |
| **Weight Update** | Feed blame + recommendations back into learning system parameters | Not designed | **Missing** |
| **Convergence Tracking** | Blame distribution stabilizing, scores improving across experiments | Not designed | **Missing** |

---

### Marcus — The Coordinator (Forward Pass)

Marcus decomposes projects into tasks, assigns them to agents, provides context, and manages the experiment lifecycle. The learning system (PatternLearner, ContextualLearner, Memory, AdaptiveDependencyInferer) represents Marcus's accumulated wisdom — the parameters that should improve over time.

**Current problem (#252):** The learning system records patterns but nothing reads them back. PatternLearner writes to `data/learned_patterns.json` and it's abandoned. Procedural memory (`self.procedural`) is entirely empty (#195). Marcus is a stateless coordinator pretending to have memory.

### Cato — The Observer (Data Layer)

Cato is the reason backpropagation is possible. Without Cato, Epictetus can only see the code. With Cato's aggregated history, Epictetus can trace any finding back to its root cause.

**What Cato already provides:**

The `ProjectHistoryAggregator` (`src/analysis/aggregator.py`) unifies 5 data sources into a single `ProjectHistory`:

| Source | What It Captures | Key Classes |
|---|---|---|
| **Conversation logs** | Task instructions agents received, agent communications | `ConversationLogger`, `Message` |
| **Agent events** | Task assignments, status changes, completions | `agent_events.log_agent_event()` |
| **Memory system** | TaskOutcomes (estimated vs actual hours, blockers, success), AgentProfiles | `Memory`, `TaskOutcome`, `AgentProfile` |
| **Project history** | Architectural decisions, artifact metadata, project snapshots | `ProjectHistoryPersistence`, `Decision`, `ArtifactMetadata`, `ProjectSnapshot` |
| **Kanban board** | Task status, comments, board state | `KanbanInterface` providers |

This produces:
- **TaskHistory** per task: instructions received, decisions made, artifacts produced, conversations, assignment times, completion times
- **AgentHistory** per agent: tasks completed, decisions made, performance profile
- **Timeline**: chronological sequence of all events across all sources

The `ProjectHistoryQuery` API (`src/analysis/query_api.py`) provides search across all of this: find tasks by status/agent/time, find decisions by impact, find artifacts by type, analyze dependency chains.

The `Events` system (`src/core/events.py`) publishes events with pub/sub, stores history (last 1000), persists to storage, and supports universal subscribers (`*` wildcard).

The `LiveExperimentMonitor` (`src/experiments/live_experiment_monitor.py`) logs everything to MLflow in real-time: agent registrations, task assignments, completions, blockers, artifacts, decisions, context requests.

**Cato's role in the loop:** Cato is the evidence chain. When Epictetus says "this code has a bug," Cato can answer: what instructions did the agent receive? What decisions did they make? What blockers did they report? What conversations happened? What was the timeline? This is the `task_instruction_envelope` that #255 says needs to be built — and the skeleton already exists in `TaskHistory.instructions_received` and `ProjectHistoryAggregator._build_task_histories()`.

### Epictetus — The Evaluator (Loss Function)

Epictetus produces structured JSON with:
- **9 dimension scores** (architecture, correctness, testing, etc.) with weights, justifications, and file:line citations
- **Agent grades** with spec adherence, ghost code detection, net contribution
- **Recommendations** scoped as `project`, `global`, or `both` — where `global` means "change Marcus itself"
- **Instruction quality assessment** (clear/ambiguous/missing/contradictory per factor)
- **Process evidence** from tmux logs and agent interrogation (optional `--session` flag)
- **Cross-agent issues**: contradictory implementations, duplicated work, integration gaps, spec drift cascades

The `global` scope recommendations and instruction quality assessment are the gradient signal. They tell you not just "the code scored 3.2/5" but "the code scored 3.2/5 *because Marcus didn't create a move/update task for the game project*" (real finding from snaky_v6).

---

### 1. PatternLearner ← Blame Attribution (system-level gradient)

**Current:** `PatternLearner._identify_failure_points()` detects high blocker rates, estimation overruns, project delays. Binary signal: "this failed."

**With the loop closed:** Blame attribution traces through Cato's timeline and tells you *why* at the system level:

```
Epictetus says: "ghost code in auth module, 3.2/5 correctness"
Cato's timeline shows: no task was created for auth testing
BlameAttributor: system_blame=0.7 — task graph gap
PatternLearner learns: project_type=backend_api → capability_template needs "auth_test"
```

That's `_analyze_workflow_patterns()` getting a causal signal it's never had: not just "this project had blockers" but "this project had blockers *because Marcus didn't decompose correctly for this project type*."

### 2. Memory Agent Profiles ← Agent Grades (agent-level gradient)

**Current:** `Memory._update_agent_profile()` uses exponential moving average on skill success rates. Binary: succeeded or failed.

**With the loop closed:** Epictetus agent grades provide magnitude. Cato's AgentHistory provides the evidence. An agent can "succeed" (task marked done in Cato) but produce 2.5/5 correctness with ghost code (per Epictetus):

```
Cato says: agent-47a3 completed task_auth at 14:32, duration 2.1 hours
Epictetus says: agent-47a3 scored 2.5/5, ghost code introduced, partial spec adherence
Memory updates: skill_success_rates["authentication"] weighted by quality, not just binary
```

The `learning_rate = 0.1` now has a continuous gradient, not a binary label.

### 3. ContextualLearner ← Instruction Quality Assessment (prompt-level gradient)

**Current:** `ContextualLearner._extract_best_practices()` learns from successful projects. No signal about *which instructions* produced quality output.

**With the loop closed:** Epictetus's instruction quality assessment rates each factor. Cato's `TaskHistory.instructions_received` shows what the agent actually got:

```
Cato shows: agent received instructions with no success criteria specified
Epictetus says: task_descriptions=clear, success_criteria=ambiguous
ContextualLearner learns: for project_type=game, explicit success criteria
  correlate with +0.8 quality improvement → adapt_template() injects success criteria
```

### 4. AdaptiveDependencyInferer ← Cross-Agent Issues (dependency gradient)

**Current:** Feature weights (`temporal_order: 0.3`, `naming_similarity: 0.2`, etc.) adapt from explicit user feedback.

**With the loop closed:** Epictetus cross-agent issues + Cato's timeline showing agents working in isolation = implicit dependency feedback:

```
Cato's timeline: Agent A and Agent B both touching auth/ between 14:00-14:30, no coordination events
Epictetus: "contradictory auth implementations, integration gap"
BlameAttributor: system_blame — dependency not specified
AdaptiveDependencyInferer: increase shared_entities weight for auth-related task pairs
```

### 5. Procedural Memory ← Global Recommendations (strategy gradient)

**Current:** `self.procedural = {"workflows": {}, "strategies": {}, "optimizations": {}}` — completely empty (#195). Nothing writes to it.

**With the loop closed:** Epictetus's `scope: "global"` recommendations + Cato's cross-project history = procedural knowledge:

```
Epictetus recommendation: {
  scope: "global",
  category: "task_templates",
  description: "Game projects need explicit move/update task"
}

Cato confirms: across 3 game experiments, task graphs consistently
  missing update/move capabilities (from ProjectHistoryAggregator)

→ procedural.workflows["game"] = {
    required_capabilities: ["design", "create", "move_update", "test"],
    evidence_count: 3,
    confidence: 0.8,  // above min_samples_for_learning threshold
    sources: ["snaky_v6", "snaky_v7", "tetris_v1"]
  }
```

After confidence exceeds the `min_samples_for_learning = 3` threshold, the workflow becomes active in task decomposition. Procedural memory earns its existence.

---

## Cato as the Missing Bridge

The key insight: **Cato is what makes blame attribution tractable.**

#255 designs a `BlameAttributor` that needs to answer questions like:
- "Was the agent told to do this?" → Cato's `TaskHistory.instructions_received`
- "Was context provided?" → Cato's `ProjectHistory.conversations` + agent events
- "Did the agent report a blocker?" → Cato's blocker events in timeline
- "What did the agent actually do and when?" → Cato's `AgentHistory` + timeline reconstruction

Without Cato's aggregated `ProjectHistory`, the BlameAttributor would need to build all of this from scratch. With Cato, it's a query:

```python
class BlameAttributor:
    def __init__(self, aggregator: ProjectHistoryAggregator):
        self.aggregator = aggregator
    
    async def attribute(self, audit_report: dict, project_id: str) -> BlameReport:
        # Cato provides the complete execution history
        history = await self.aggregator.aggregate_project(project_id)
        
        for issue in audit_report["issues"]["critical"]:
            # Find what the agent was told (from Cato)
            task_history = self.find_task(history.tasks, issue)
            instructions = task_history.instructions_received
            decisions = task_history.decisions_made
            timeline = [e for e in history.timeline if e.task_id == task_history.task_id]
            
            # Now we can attribute: was this a system problem or agent problem?
            blame = self.classify(issue, instructions, decisions, timeline)
```

The `ProjectHistoryQuery` API means Epictetus could also query Cato directly during its audit phases — not just reading code, but searching the execution history for patterns that explain the code.

---

## The Convergence Signal

Track per-experiment metrics to verify the system is actually learning:

```json
{
  "experiment": "snaky_v6",
  "score": 3.12,
  "blame": {"system": 0.60, "agent": 0.35, "prompt": 0.05},
  "patterns_applied": 0,
  "cato_data_quality": {"timeline_events": 142, "instructions_captured": 4, "decisions_logged": 8},
  "optimizer_version": null
}
```

Over time:
```
Exp 1: system=60%, agent=35%, score=3.1 — no patterns, Cato observing
Exp 2: system=45%, agent=40%, score=3.5 — system learning from Exp 1 audit
Exp 3: system=30%, agent=50%, score=3.8 — prompt fix, agents now bottleneck
Exp 4: system=25%, agent=35%, score=4.2 — converging, Cato confirms improvements
```

Cato provides the ground truth for convergence: are agents actually receiving better instructions? Are task graphs actually more complete? Are dependency specifications actually preventing integration gaps? The learning system says it updated parameters — Cato verifies the parameters produced different behavior.

---

### Phase 0: Prerequisites
- [ ] Wire existing learning system into production (#252)
- [ ] Complete procedural memory tier (#195)
- [ ] Enrich Cato's traceability: ensure `TaskHistory.instructions_received` captures the full instruction envelope, not just task description (#255 Phase 1)
- [ ] Agent-signed git commits for code-level attribution (#121)
- [ ] Task graph coverage logging after decomposition (#255 Part 1c)

### Phase 1: The Adapter (Epictetus JSON → Learning APIs via Cato)
- [ ] Build `AuditIngester` that reads Epictetus JSON and Cato's `ProjectHistory` together
- [ ] `PatternLearner.learn_from_project()` accepts optional `audit_report` parameter for quality-enriched learning
- [ ] `Memory._update_agent_profile()` accepts quality score (from Epictetus) alongside binary success (from Cato)
- [ ] `ContextualLearner` correlates instruction quality (Epictetus) with actual instructions delivered (Cato)
- [ ] Store audit-derived patterns alongside existing patterns

### Phase 2: The Blame Engine (#255 Phase 2)
- [ ] Build `BlameAttributor` that takes Epictetus JSON + Cato's `ProjectHistoryAggregator` output
- [ ] Use Cato's `TaskHistory` to answer "was the agent told to do this?"
- [ ] Use Cato's timeline to answer "what happened and when?"
- [ ] Use Cato's `AgentHistory` to answer "what did this agent do across all tasks?"
- [ ] Implement blame classification rules (system/agent/prompt allocation)
- [ ] Add `blame_attribution` section to Epictetus report schema

### Phase 3: The Optimizer (#255 Phase 3)
- [ ] `TaskDecompositionOptimizer` — pre-flight task graph validation against learned capability templates (informed by Cato's cross-project task patterns)
- [ ] `PromptOptimizer` — detect recurring CLAUDE.md tensions using Cato's conversation logs + Epictetus findings
- [ ] `ContextOptimizer` — improve context using Cato's data on what context was provided vs. what was needed
- [ ] Write procedural memory entries from global recommendations

### Phase 4: Convergence Tracking
- [ ] Store blame distribution + score + Cato data quality metrics per experiment
- [ ] Track blame trajectory across experiments per project type
- [ ] Use Cato to verify that parameter updates actually changed behavior (not just recorded)
- [ ] Alert when same recommendation appears in 3+ audits (not being applied despite being learned)

### Phase 5: Cross-Project Learning (#195 Phase 4)
- [ ] `MemoryExporter` serializes learned knowledge (agent profiles, patterns, procedural entries, audit-derived strategies) to portable format
- [ ] `MemoryImporter` loads prior knowledge with decay factors
- [ ] Cato's cross-project `ProjectHistoryAggregator` provides the validation: does imported knowledge actually improve outcomes?

---

## The Stoic Architecture

The naming was always intentional:

- **Marcus Aurelius** — the emperor who coordinates, who acts in the world. Marcus decomposes, assigns, orchestrates. The forward pass.
- **Cato the Elder** — the censor, the observer, the one who watches and records. Cato aggregates every event, decision, conversation, and artifact into a unified history. The data layer.
- **Epictetus** — the philosopher who examines, who asks "did you act wisely?" Epictetus evaluates the work against a rigorous rubric and identifies what went wrong. The loss function.

Together they embody the Stoic practice: act (Marcus), observe (Cato), examine (Epictetus), improve (the learning loop). The philosophy is now the architecture.

No other multi-agent framework has this. AutoGen, CrewAI, and LangGraph are stateless — they act but never observe, examine, or improve. Marcus with this loop closed is a **self-improving coordination system** where each experiment makes the next one measurably better, and the evidence chain (Cato) proves it.

---

## Related Issues

- #195 — Cognitive Memory System: Procedural Memory and Cross-Project Learning (the empty `procedural` dict)
- #255 — Backpropagation-style blame attribution (the backward pass design, traceability infrastructure)
- #252 — Activate learning feedback loop (prerequisite: wire PatternLearner into production)
- #176 — Close the Learning Loop (the original observation that learning → storage is a dead end)
- #121 — Git traceability (agent-signed commits for code-level blame attribution)
- #196 — Context access logging (instruction envelopes — partially covered by Cato's TaskHistory)
- #122 — Artifact system workflow validation (spec-vs-impl checking)
- #228 — Audit log missing duration_ms (Cato observability gap)

---

*Marcus acts. Cato watches. Epictetus judges. The learning system remembers. Next experiment is better. Repeat until convergence.*

**Comment by lwgray** [comment]
Related: #267 — topology-aware task decomposition addresses the planning failures that Epictetus audits detect (single-author products due to feature-based decomposition of tightly coupled code).
